lib/internal/util.js

```diff
@@ -789,6 +794,24 @@ function setupCoverageHooks(dir) {
   return coverageDirectory;
 }
 
+// This would better be placed in internal/worker/io.js, but that doesn't work
+// because Buffer needs this and that would introduce a cyclic dependency.
+// Mark the object as untransferable. If object occurs in the transfer list of
+// a port.postMessage() call, it is ignored.
+function markAsUntransferable(obj) {
```

Comment:
The terminology is pretty confusing. Why is this called "Untransferable" when what it does is to mark it as "Clonable"?

From what I can tell, the three adjectives mean (and let me know if this is wrong):

* Transferable: cannot be cloned but can be transferred
* Clonable: can be cloned but cannot be transferred
* Unclonable: cannot be cloned and cannot be transferred

But what about objects that can be transferred, and can also be cloned? Examples include `ArrayBuffer`.

If I misunderstood and Transferable objects can also be cloned, then what about objects that cannot be cloned, but can be transferred? Examples include `MessagePort`.

Reply:
Yeah, the terms here are confusing when it is compared to WebIDL's definition. The existing implementation doesn't strictly align with the WebIDL's [Serializable](https://html.spec.whatwg.org/multipage/structured-data.html#serializable-objects) and [Transferable](https://html.spec.whatwg.org/multipage/structured-data.html#transferable-objects) definition, e.g. the `JSTransferable` attempts to clone the object when a transfer is required ([link](https://github.com/nodejs/node/blob/c94be4125bfbc68e4cd9cbec27676347a314997d/src/node_messaging.cc#L1229)).

I agree that we need to distinguish between Transferable and Serializable. However, the term "serialize" is currently being used by [snapshotable](https://github.com/nodejs/node/blob/c94be4125bfbc68e4cd9cbec27676347a314997d/src/node_snapshotable.h#L106). It's not a public API and can be renamed anyways. I'm going to apply the alignment in a separate PR.

Reply:
I think it's fine to call Web IDL's Serializable `Clonable` in the context of Node.js. But it seems like what we need is two bits, one for Clonability, one for Transferability, rather than a tri-state.